File: MVVM/Converters/BrightnessConverter.vb
Imports System.Globalization
Imports System.Windows.Data
Imports System.Windows.Media

Namespace MVVM.Converters

    Public Class BrightnessConverter
        Inherits DependencyObject
        Implements IValueConverter

        Public Function Convert(ByVal Value As Object, ByVal TargetType As Type, ByVal Parameter As Object, ByVal Culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim B = TryCast(Value, SolidColorBrush)
            Dim C As Color
            If B IsNot Nothing Then
                C = B.Color
            Else
                C = CType(Value, Color)
            End If

            Dim P = 0.0
            If Parameter IsNot Nothing Then
                P = CType(Parameter, Double)
            End If

            C = Me.Convert(C, P)

            If TargetType = GetType(Color) Then
                Return C
            End If
            Return New SolidColorBrush(C)
        End Function

        Public Function ConvertBack(ByVal Value As Object, ByVal TargetType As Type, ByVal Parameter As Object, ByVal Culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Dim B = TryCast(Value, SolidColorBrush)
            Dim C As Color
            If B IsNot Nothing Then
                C = B.Color
            Else
                C = CType(Value, Color)
            End If

            Dim P = 0.0
            If Parameter IsNot Nothing Then
                P = CType(Parameter, Double)
            End If

            C = Me.Convert(C, -P)

            If TargetType = GetType(Color) Then
                Return C
            End If
            Return New SolidColorBrush(C)
        End Function

        Private Shared Function GetByte(ByVal N As Integer) As Byte
            If N < 0 Then
                N = 0
            End If
            If N > 255 Then
                N = 255
            End If
            Return CByte(N)
        End Function

        Public Function Convert(ByVal C As Color, Optional ByVal Parameter As Double = 0.0) As Color
            Dim P = CInt(Parameter * 255 * Me.Coeff)
            Return Color.FromArgb(C.A, GetByte(C.R + P), GetByte(C.G + P), GetByte(C.B + P))
        End Function

        Private Sub CalculateCoeff()
            Dim Background = Me.Background
            Dim Foreground = Me.Foreground
            Dim BG = (Background.ScR + Background.ScG + Background.ScB) / 3
            Dim FG = (Foreground.ScR + Foreground.ScG + Foreground.ScB) / 3

            Me.Coeff = BG - FG
        End Sub

#Region "Background Property"
        Public Shared ReadOnly BackgroundProperty As DependencyProperty = DependencyProperty.Register("Background", GetType(Color), GetType(BrightnessConverter), New PropertyMetadata(Colors.White, AddressOf Background_Changed))

        Private Shared Sub Background_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, BrightnessConverter)
            Self.CalculateCoeff()
        End Sub

        Public Property Background As Color
            Get
                Return DirectCast(Me.GetValue(BackgroundProperty), Color)
            End Get
            Set(value As Color)
                Me.SetValue(BackgroundProperty, value)
            End Set
        End Property
#End Region

#Region "Foreground Property"
        Public Shared ReadOnly ForegroundProperty As DependencyProperty = DependencyProperty.Register("Foreground", GetType(Color), GetType(BrightnessConverter), New PropertyMetadata(Colors.Black, AddressOf Foreground_Changed))

        Private Shared Sub Foreground_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, BrightnessConverter)
            Self.CalculateCoeff()
        End Sub

        Public Property Foreground As Color
            Get
                Return DirectCast(Me.GetValue(ForegroundProperty), Color)
            End Get
            Set(value As Color)
                Me.SetValue(ForegroundProperty, value)
            End Set
        End Property
#End Region

        Private Coeff As Double = 1

    End Class

End Namespace

File: WPF/MVVM/Converters/BrightnessConverter.cs
using System;
using System.Globalization;
using System.Windows;
using System.Windows.Data;
using System.Windows.Media;

namespace Ks.Common.MVVM.Converters
{
    public class BrightnessConverter : DependencyObject, IValueConverter
    {
        public object Convert(object Value, Type TargetType, object Parameter, CultureInfo Culture)
        {
            var B = Value as SolidColorBrush;
            Color C;
            if (B != null)
            {
                C = B.Color;
            }
            else
            {
                C = (Color) Value;
            }

            var P = 0.0;
            if (Parameter != null)
            {
                P = System.Convert.ToDouble(Parameter);
            }

            C = this.Convert(C, P);

            if (TargetType == typeof(Color))
            {
                return C;
            }

            return new SolidColorBrush(C);
        }

        public object ConvertBack(object Value, Type TargetType, object Parameter, CultureInfo Culture)
        {
            var B = Value as SolidColorBrush;
            Color C;
            if (B != null)
            {
                C = B.Color;
            }
            else
            {
                C = (Color) Value;
            }

            var P = 0.0;
            if (Parameter != null)
            {
                P = System.Convert.ToDouble(Parameter);
            }

            C = this.Convert(C, -P);

            if (TargetType == typeof(Color))
            {
                return C;
            }

            return new SolidColorBrush(C);
        }

        private static byte GetByte(int N)
        {
            if (N < 0)
            {
                N = 0;
            }

            if (N > 255)
            {
                N = 255;
            }

            return (byte) N;
        }

        public Color Convert(Color C, double Parameter = 0.0)
        {
            var P = System.Convert.ToInt32(Parameter * 255 * this.Coeff);
            return Color.FromArgb(C.A, GetByte(C.R + P), GetByte(C.G + P), GetByte(C.B + P));
        }

        private void CalculateCoeff()
        {
            var Background = this.Background;
            var Foreground = this.Foreground;
            var BG = (Background.ScR + Background.ScG + Background.ScB) / 3;
            var FG = (Foreground.ScR + Foreground.ScG + Foreground.ScB) / 3;

            this.Coeff = BG - FG;
        }

        public static readonly DependencyProperty BackgroundProperty = DependencyProperty.Register("Background", typeof(Color), typeof(BrightnessConverter), new PropertyMetadata(Colors.White, Background_Changed));

        private static void Background_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (BrightnessConverter) D;
            Self.CalculateCoeff();
        }

        public Color Background
        {
            get
            {
                return (Color) this.GetValue(BackgroundProperty);
            }
            set
            {
                this.SetValue(BackgroundProperty, value);
            }
        }

        public static readonly DependencyProperty ForegroundProperty = DependencyProperty.Register("Foreground", typeof(Color), typeof(BrightnessConverter), new PropertyMetadata(Colors.Black, Foreground_Changed));

        private static void Foreground_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (BrightnessConverter) D;
            Self.CalculateCoeff();
        }

        public Color Foreground
        {
            get
            {
                return (Color) this.GetValue(ForegroundProperty);
            }
            set
            {
                this.SetValue(ForegroundProperty, value);
            }
        }

        private double Coeff = 1;
    }
}
